File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int num_node = 0;
        
        ListNode* ptr = head;
        while(ptr != nullptr) {
            ptr = ptr->next;
            num_node++;
        }
        
        ptr = head;
        for (int i = 0; i < num_node - n - 1; i++) {
            ptr = ptr->next;
        }
        
        if (num_node == n) {
            return head->next;
        }
        ListNode* to_be_deleted = ptr->next;
        ListNode* next = to_be_deleted->next;
        ptr->next = next;
        delete to_be_deleted;
        
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Two pointer method
        # Distane between the two pointer will always be n
        dummy = ListNode(0, head)
        left = dummy
        right = head
        
        # Shifting the right pointer
        while n > 0:
            right = right.next
            n -= 1
        
        # Shifting both pointer
        while right:
            right = right.next
            left = left.next
            
        # Delete 
        left.next = left.next.next
        return dummy.next